helios-services/src/main/java/com/spotify/helios/agent/HealthCheckerFactory.java

```diff
@@ -55,19 +64,76 @@ public static HealthChecker create(final TaskConfig taskConfig, final DockerClie
     throw new IllegalArgumentException("Unknown healthCheck type");
   }
 
-  private static class ExecHealthChecker implements HealthChecker {
+  static class ExecHealthChecker implements HealthChecker {
+
+    private static final Logger log = LoggerFactory.getLogger(ExecHealthChecker.class);
 
     private final ExecHealthCheck healthCheck;
     private final DockerClient docker;
 
-    private ExecHealthChecker(final ExecHealthCheck healthCheck, final DockerClient docker) {
+    ExecHealthChecker(final ExecHealthCheck healthCheck, final DockerClient docker) {
       this.healthCheck = healthCheck;
       this.docker = docker;
     }
 
     @Override
     public boolean check(final String containerId) {
-      throw new UnsupportedOperationException("not yet supported");
+      // Make sure we are on a docker version that supports exec health checks
+      if (!compatibleDockerVersion(docker)) {
+        throw new UnsupportedOperationException(
+            "docker exec healthcheck is not supported on your docker version");
+      }
+
+      try {
+        final List<String> cmd = healthCheck.getCommand();
+        final String execId = docker.execCreate(containerId, cmd.toArray(new String[cmd.size()]),
+                                                DockerClient.ExecParameter.STDOUT,
+                                                DockerClient.ExecParameter.STDERR);
+
+        final String output;
+        try (LogStream stream = docker.execStart(execId)) {
+          output = stream.readFully();
+        }
+
+        final int exitCode = docker.execInspect(execId).exitCode();
+        if (exitCode != 0) {
+          log.info("healthcheck failed with exit code {}. output {}", exitCode, output);
+          return false;
+        }
+
+        return true;
+      } catch (DockerException e) {
+        return false;
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        return false;
+      }
+    }
+
+    private static boolean compatibleDockerVersion(final DockerClient docker) {
+      final String executionDriver;
```

Comment:
One thing i just realized, I believe we already have the Docker version in the agent state somewhere. That's what we get from `helios hosts`. So we should use that instead of making a new Docker request.